src/cmd/ksh93/sh/macro.c

```diff
@@ -1792,10 +1798,16 @@ static_fn bool varsub(Mac_t *mp) {
                 mp->atmode = mode == '@';
                 mp->pattern = oldpat;
             } else if (d) {
-                if (mp->sp) {
-                    sfputc(mp->sp, d);
+                Sfio_t *sfio_ptr = (mp->sp) ? mp->sp : stkp;
+
+                // Handle multi-byte characters being used for the internal
+                // field separator (IFS).
+                if (ifs_bytes) {
```

Comment:
Why introduce `ifs_bytes` rather than just test `mode == '@'` here? The existing code is already schizophrenic which I don't expect you to fix. For example, this definition of a member of `Mac_t`:

```
unsigned char ifs;        // first char of IFS
```

First, it isn't a non-ASCII char and second, even within the context of ASCII why is this var an `unsigned char` where the underlying var is a simple `char` which may or may not be signed.

Reply:
Too, I think it is more important to make this test very clear that it is really testing whether a "@" expansion or not it being handled rather than doing so indirectly via the variable you introduced. Long term this hideous function needs to be refactored. And the entire code base should consistently use wide chars or use UTF-8 internally. But I expect to be dead before that happens.

Reply:
I suspect I was being overly conservative here - I'm a little afraid to reuse the defined variables in the method since it's not always easy to tell what unexpected side-effects an additional usage might introduce.

This way is definitely simpler and nixes the extra variable, so I've gone ahead and migrated to using this approach.